### ISSUE TITLE
boson radiometric 

### ISSUE THREAD

**Issue by bakas90** [LJMUAstroecology/flirpy#87]
I'm using a Boson R camera.
I am getting raw data using flirpy.

However, I'm not exactly sure how to get the metadata.

I want to get the radiometric value.
I also want to create a SEQ file.

I would be grateful if you could tell me how.

**Comment by jveitchmichaelis** [comment]
Hi, 

> I want to get the radiometric value.

Flirpy was written before the R variant was released and unfortunately I don't have a radiometric camera to test with.

Do you know what format the raw data from the camera is returned as? It should be straightforward to update the library to include radiometric support. Documentation is [here](https://flir.netx.net/file/asset/46046/original/attachment)

I'd guess we need to check if TLinear mode is enabled at least, and there seem to be options to control what units are output.

> I also want to create a SEQ file.

SEQ files are a sequence of FFF files. In theory you could construct these manually, but this isn't something that Flirpy supports - it's mostly used in the opposite direction (getting data out of SEQ files).

Is there a reason you want an SEQ file?

**Comment by jveitchmichaelis** [comment]
Assuming you've created a camera object, give this a go. This should change the video mode to TLinear.

```
function_id = 0x000E000E #sysctrlSetUsbVideoIR16Mode
command = struct.pack(">H", 2) # FLR_SYSCTRL_USBIR16_MODE_TLINEAR = 2
res = camera._send_packet(function_id, data=command)
res = camera._decode_packet(res)
```

This should put your camera into TLinear mode and then multiply your raw values by 0.01 to get to Kelvin.

**Comment by jveitchmichaelis** [comment]
@bakas90 I see the thumbs up, any luck with this? :)

**Comment by olfk1234** [comment]
@jveitchmichaelis Hello! I tried the code above and I got a warning stating "WARNING:flirpy.camera.boson:Boson response: bad command ID" when running the decode packet function. Is there another workaround to convert the Boson into TLinear mode or to get the temperatures using flirpy?

**Comment by jveitchmichaelis** [comment]
If you go to the receive function and print/dump the return from the camera, what do you get?

I don't have a radiometric Boson so I'm unable to check this myself.

**Comment by olfk1234** [comment]
@jveitchmichaelis After doing `print(camera.receive())` after the code that you mentioned before, I am getting `b' '`

**Comment by jveitchmichaelis** [comment]
Try with `0x000E000D`:

<img width="907" alt="image" src="https://github.com/LJMUAstroecology/flirpy/assets/3159591/54eb5ec5-adff-4fc4-b1c2-cbf8120e16f5">

Looks like a typo with the previous function code (that was get, not set).

**Comment by millercommamatt** [comment]
I'm just diving into flirpy for my radiometric Boson. I requested the USB Vidieo IR Mode as show below. The returned status value was 2 which should be T-Linear mode. Grabbing an image resulting the an unsigned 8-bit numpy array which doesn't seem correct to me. I was expecting a uint16 array.

I'm sure I'm missing something simple, but this is my first time both with this library and more importantly, a Boson camera.

```
>>> res=camera._send_packet(0x000E000E)
>>> res=camera._decode_packet(res,receive_size=4)
>>> res
b'\x00\x00\x00\x02'
>>> image=camera.grab()
>>> image
array([[ 93, 117,  93, ..., 117,  86, 117]],
      shape=(1, 655360), dtype=uint8)

>>> res=camera._send_packet(0x003E0002)  # TLinearGetControl()
>>> res=camera._decode_packet(res,receive_size=4)
>>> res
b'\x00\x00\x00\x01'  # 1 = FLR_ENABLE

```

**Comment by jveitchmichaelis** [comment]
Yeah though you shouldn't need to use the low level functions to do this. Interesting that grab is not returning a 16 bit array though, it should handle that conversion for you...

**Comment by millercommamatt** [comment]
I'll dig into why OpenCV isn't returning the expected data structure. Issue #105 reported the same phenomenon. Let me know if you have troubleshooting suggestions. I've not tangled with OpenCV before.

I'm using a Radiometric Boson 640.

Thanks for this library. It saved me a ton a work. I might send some PRs to expand some of the functionality for the Boson class.

**Comment by jveitchmichaelis** [comment]
Yeah it looks like a regression in OpenCV so you could try and bisect to figure out where it was introduced and we could raise an issue directly.

Otherwise I would really welcome PRs for radiometric functionality as this is something I don't have the ability to test at the moment.

Thanks!

**Comment by jveitchmichaelis** [comment]
By the way you can also try casting the array directly using `view`: https://numpy.org/doc/2.2/reference/generated/numpy.ndarray.view.html

**Comment by millercommamatt** [comment]
Perhaps this is the origin of the apparent change in OpenCV behavior?
https://github.com/opencv/opencv/pull/18498/files

Note the changes in output data types for `case V4L2_PIX_FMT_Y16:`...

**Comment by jveitchmichaelis** [comment]
Interesting. I'll have a look. So are you able to confirm whether a build either side of that commit passes/fails?

What's unclear to me without digging is why we're getting a flat array back and not just some messed up values (like if the endianness is backwards). Grab should return a 2D image if the VideoCapture object has some idea of width/height. What happens if you query the capture properties with a build of OpenCV that gives you the wrong shape output?

**Comment by jveitchmichaelis** [comment]
@millercommamatt could you confirm this fix please?

https://github.com/opencv/opencv/pull/27193
